WebSocket 예외 응답 구조 표준화

yxxjxxeee opened the pull request:
## #️⃣ 연관된 이슈

- #47 

## 📝 작업 내용

- WebSocket에서 발생하는 예외를 공통된 응답 구조로 통일했습니다.
1. WsException이 문자열이 아닌 표준 예외 객체를 던지도록 수정
2. SocketExceptionFilter에서 예외 응답을 표준 포맷(status, code, errors)으로 통일
3. Gateway 메서드에서 발생하는 비즈니스 예외를 표준 구조로 변환

## 💬 리뷰 요청 사항

- 웹소켓 응답 구조도 status, code, errors 형태로 통일하는 것이 적절한지 궁금합니다.

yxxjxxeee committed:
refactor: (#47) socket 예외 처리 통일

yxxjxxeee committed:
refactor: (#47) WebSocket 예외 구조 수정

yxxjxxeee committed:
refactor: (#47) ErrorResponseBuilder 사용으로 예외 처리 로직 중복 제거

kimjiwonpg98 commented on src/chat/chat.gateway.ts:
```diff
@@ -69,28 +101,37 @@ export class ChatGateway implements OnGatewayConnection, OnGatewayDisconnect {
+        code: isTokenExpired
+          ? CHAT_ERROR_CODES.EXPIRED_ACCESS_TOKEN
+          : CHAT_ERROR_CODES.INVALID_ACCESS_TOKEN,
+        errors: [],
```
표준화를 한다면 유틸 함수를 만들어서 오브젝트를 관리하는 것도 좋아보이네요

yxxjxxeee replied:
좋은 것 같습니다 !!

yxxjxxeee committed:
refactor: (#47) 유틸 도입으로 웹소켓 에러 생성 표준화

Kimsoo0119 commented on src/common/filters/global-exception.filter.ts:
```diff
@@ -0,0 +1,34 @@
+    if (status >= 500) {
+      const logMessage =
+        exception instanceof Error ? exception.stack : JSON.stringify(exception, null, 2);
+      this.logger.error('Unhandled exception', logMessage);
```
로깅할 때 스택만 남기면 나중에 어떤 요청에서 터졌는지 찾기 힘들 것 같아요. request.method랑 request.url 정도는 같이 남기면 디버깅할 때 편할 것 같습니다. 
```ts
this.logger.error(`[${request.method} ${request.url}] Unhandled exception`, logMessage);
```

Kimsoo0119 replied:
추가로 4xx 쪽 로깅은 따로 안하시는 이유가 있을까요?

yxxjxxeee replied:
좋은 의견 감사합니다!
말씀 주신 것처럼 stack 정보만 남기면 어떤 요청에서 발생한 에러인지 추적하기 어렵기 때문에
request.method, request.url 같은 요청 컨텍스트를 함께 로깅하는 것이 디버깅에 훨씬 도움이 될 것 같네요! 

다만 이번 PR에서는 에러 응답 포맷과 예외 처리 흐름을 정리하는 것까지를 범위로 가져가고
실제 로깅 전략은 winston + CloudWatch 연동과 함께 다음 이슈에서 정리할 예정입니다.

Kimsoo0119 commented on src/common/builders/error-response.builder.ts:
```diff
@@ -0,0 +1,56 @@
+      : [];
+
+    return { status, code, errors };
+  }
```
static 메서드 하나만 있는 구조라서 함수로 사용해도 괜찮지 않을까 합니다

rrgks6221 commented on src/common/utils/ws-error.util.ts:
```diff
@@ -0,0 +1,24 @@
+import { ErrorDetail, StandardErrorResponse } from '../builders/error-response.builder';
+import { WsException } from '@nestjs/websockets';
+
+export class WsError {
```
개인적으로는 생성자만으로 충분하지 않을까 해요 status에 따라서 계속 메서드가 추가되는 유지보수 비용도 있고 개발자가 어떤 statuscode에 해당하는 메서드인지를 찾아야하는 수고도 있을거같아요
추가로 websocket에는 http status code처럼 표준화된 에러 응답 코드는 없고 소켓 연결이 종료됐을 때 사용하는 close code 정도만 있는거로 알고 있어요

https://developer.mozilla.org/en-US/docs/Web/API/CloseEvent/code

yxxjxxeee replied:
좋은 의견 감사합니다! 리뷰 남겨주신 것처럼 status별 메서드가 늘어나는 구조보다는 단일 생성 방식을 사용하는 것이 유지보수 측면에서 더 낫다고 생각해서 해당 방향으로 수정했습니다
WebSocket에는 표준 status 개념이 없다는 점은 인지하고 있지만 현재 프로젝트에서는 REST API와 WebSocket 양쪽에서 에러 응답 형식을 통일하는 것을 우선 목표로 두고 있어서 WebSocket 쪽에서도 status, code, errors 형태를 유지하도록 구현했습니다!

rrgks6221 commented on src/rooms/rooms.gateway.ts:
```diff
@@ -19,30 +22,28 @@ export class RoomsGateway implements IRoomsEventPublisher {
-      this.logger.error(`Failed to emit room created: ${room.id}`, err);
+      this.logger.error(`Failed to emit event: ${event}`, err as Error);
+
+      throw WsError.internalServerError('ROOM_EVENT_FAILED');
```
미리 선언해놓은 상수 사용이 안됐네요

```suggestion
      throw WsError.internalServerError(ROOM_ERROR_CODES.ROOM_EVENT_FAILED);
```

yxxjxxeee committed:
refactor: (#47) WebSocket 에러 처리 구조 개선

- WsError 클래스를 wsError() 유틸 함수로 변경
- ErrorResponseBuilder static 메서드를 buildErrorResponse() 함수로 변경
- ROOM_ERROR_CODES 상수 사용으로 하드코딩 제거

yxxjxxeee committed:
Merge branch 'develop' of https://github.com/modern-agile-team/10term-pollycasso-back into refactor/#47/ws-error-standardization
